flux_led/protocol.py

```diff
@@ -0,0 +1,234 @@
+        """The name of the protocol."""
+
+    @property
+    @abstractmethod
+    def state_response_names(self):
+        """The names of the values in the state response."""
+
+    @property
+    @abstractmethod
+    def set_command_names(self):
+        """The names of the values in the set command."""
+
+    @property
+    def state_response_length(self):
+        """The length of the query response."""
+        return len(self.state_response_names)
+
+    @abstractmethod
+    def construct_message(self, raw_bytes):
+        """Original protocol uses no checksum."""
+
+
+class ProtocolLEDENETOriginal(ProtocolBase):
+    """The original LEDENET protocol with no checksums."""
+
+    @property
+    def name(self):
+        """The name of the protocol."""
+        return PROTOCOL_LEDENET_ORIGINAL
+
+    def is_valid_state_response(self, raw_state):
+        """Check if a state response is valid."""
+        return len(raw_state) == self.state_response_length and raw_state[1] == 0x01
+
+    def construct_state_query(self):
+        """The bytes to send for a query request."""
+        return self.construct_message(bytearray([0xEF, 0x01, 0x77]))
+
+    def construct_state_change(self, turn_on):
+        """The bytes to send for a state change request."""
+        return self.construct_message(
+            bytearray([0xCC, self.on_byte if turn_on else self.off_byte, 0x33])
+        )
+
+    def construct_message(self, raw_bytes):
+        """Original protocol uses no checksum."""
+        return raw_bytes
+
+    # typical response:
+    # pos  0  1  2  3  4  5  6  7  8  9 10
+    #    66 01 24 39 21 0a ff 00 00 01 99
+    #     |  |  |  |  |  |  |  |  |  |  |
+    #     |  |  |  |  |  |  |  |  |  |  checksum
+    #     |  |  |  |  |  |  |  |  |  warmwhite
+    #     |  |  |  |  |  |  |  |  blue
+    #     |  |  |  |  |  |  |  green
+    #     |  |  |  |  |  |  red
+    #     |  |  |  |  |  speed: 0f = highest f0 is lowest
+    #     |  |  |  |  <don't know yet>
+    #     |  |  |  preset pattern
+    #     |  |  off(24)/on(23)
+    #     |  type
+    #     msg head
+    #
+    @property
+    def state_response_names(self):
+        """The names of the values in the state response."""
+        return [
+            "head",
+            "type",
+            "power_state",
+            "preset_pattern",
+            "mode",
+            "speed",
+            "red",
+            "green",
+            "blue",
+            "warm_white",
+            "check_sum",
+        ]
+
+    @property
+    def set_command_names(self):
+        """The names of the values in the set command."""
+        return [
+            "head",
+            "red",
+            "green",
+            "blue",
+            "terminator",
+        ]
+
+
+class ProtocolLEDENET8Byte(ProtocolBase):
+    """The newer LEDENET protocol with checksums that uses 8 bytes to set state."""
+
+    def is_valid_state_response(self, raw_state):
+        """Check if a state response is valid."""
```

Comment:
In another PR, it would be good to validate the checksum as well.

Since these devices reply to every client regardless of who sent the message, this validation isn't strong enough since we could happen to get a message with a RGBWW value of 0x81 that perfectly lines up and we think we have a good state message